controllers/cloud.redhat.com/clowdapp_reconciliation.go

```diff
@@ -0,0 +1,381 @@
+				r.log.Info("Cloud not finalize", "err", finalizeErr)
+				return ctrl.Result{}, finalizeErr
+			}
+
+			controllerutil.RemoveFinalizer(r.app, appFinalizer)
+			removeFinalizeErr := r.client.Update(*r.ctx, r.app)
+			if removeFinalizeErr != nil {
+				r.log.Info("Cloud not remove finalizer", "err", removeFinalizeErr)
+				return ctrl.Result{}, removeFinalizeErr
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) finalizeApp() error {
+	delete(managedApps, r.app.GetIdent())
+	managedAppsMetric.Set(float64(len(managedApps)))
+
+	delete(presentApps, r.app.GetIdent())
+	presentAppsMetric.Set(float64(len(presentApps)))
+
+	r.log.Info("Successfully finalized ClowdApp")
+	return nil
+}
+
+func (r *ClowdAppReconciliation) addFinalizer() (ctrl.Result, error) {
+	if !contains(r.app.GetFinalizers(), appFinalizer) {
+		if addFinalizeErr := r.addFinalizerImplementation(); addFinalizeErr != nil {
+			r.log.Info("Cloud not add finalizer", "err", addFinalizeErr)
+			return ctrl.Result{}, addFinalizeErr
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) addFinalizerImplementation() error {
+	r.log.Info("Adding Finalizer for the ClowdApp")
+	controllerutil.AddFinalizer(r.app, appFinalizer)
+
+	// Update CR
+	err := r.client.Update(context.TODO(), r.app)
+	if err != nil {
+		r.log.Error(err, "Failed to update ClowdApp with finalizer")
+		return err
+	}
+	return nil
+}
+
+func (r *ClowdAppReconciliation) isEnvLocked() (ctrl.Result, error) {
+	if ReadEnv() == r.app.Spec.EnvName {
+		r.recorder.Eventf(r.app, "Warning", "ClowdEnvLocked", "Clowder Environment [%s] is locked", r.app.Spec.EnvName)
+		r.log.Info("Env currently being reconciled")
+		return ctrl.Result{Requeue: true}, errors.New(SKIPRECONCILE)
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) isAppDisabled() (ctrl.Result, error) {
+	if r.app.Spec.Disabled {
+		r.log.Info("Reconciliation aborted - set to be disabled")
+		return ctrl.Result{}, errors.New(SKIPRECONCILE)
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) getClowdEnv() (ctrl.Result, error) {
+	updatedContext := context.WithValue(*r.ctx, errors.ClowdKey("obj"), r.app)
+	r.ctx = &updatedContext
+	r.env = &crd.ClowdEnvironment{}
+
+	if getEnvErr := r.client.Get(*r.ctx, types.NamespacedName{Name: r.app.Spec.EnvName}, r.env); getEnvErr != nil {
+		r.log.Info("ClowdEnv missing", "err", getEnvErr)
+		r.recorder.Eventf(r.app, "Warning", "ClowdEnvMissing", "Clowder Environment [%s] is missing", r.app.Spec.EnvName)
+		if setClowdStatusErr := SetClowdAppConditions(*r.ctx, r.client, r.app, crd.ReconciliationFailed, getEnvErr); setClowdStatusErr != nil {
+			r.log.Info("Set status error", "err", setClowdStatusErr)
+			return ctrl.Result{Requeue: true}, setClowdStatusErr
+		}
+		return ctrl.Result{}, getEnvErr
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) isEnvCoreNamespaceDeleted() (ctrl.Result, error) {
```

Comment:
Do we need `Core` in here?

Reply:
That was some cargo cult naming on my part. I didn't know if the fact that we were dealing with a core namespace was in some way significant. Judging from context I'm assuming it isn't :) I'll remove it..